File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        return dfs(root).second;
    }
    
    pair<int, TreeNode*> dfs(TreeNode* root) {
        if(!root) return {-1, nullptr};
        auto l = dfs(root->left);
        auto r = dfs(root->right);
        int dl = l.first; int dr = r.first;
        return {max(dl, dr)+1, dl == dr ? root : dl > dr ? l.second : r.second};
    }
};

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var subtreeWithAllDeepest = function(root) {
    return dfs(root).node;
};

function dfs(root) {
    if(root == null) return {depth: -1, node: root};
    let left = dfs(root.left);
    let right = dfs(root.right);
    if(left.depth == right.depth) return {depth: left.depth+1, node: root};
    if(left.depth > right.depth) return {depth: left.depth+1, node: left.node};
    else return {depth: right.depth+1, node: right.node};
}